### PR TITLE
Fix to issue #38 (AlexNet)

### LINKED ISSUE #38: Build a Keras CNN model using Alexnet Architecture
Create a keras Convolutional Neural Network on Alexnet arch. Create model checkpoints in the code.

### PR DESCRIPTION
**Pull request by AdityaSingh17**
This pull request solves issue #38 
Files added:  CropAI_AlexNet.ipynb

### CODE REVIEW

**Review by gshanbhag525 — CHANGES_REQUESTED** [review]
I need you to do following changes. 

- Since plantdisease.zip after extraction becomes PlantVillage. Change line

`directory_root = 'PlantVillage'`

- `for plant_disease_folder in plant_disease_folder_list:
            print(f"[INFO] Processing {plant_disease_folder} ...")
            plant_disease_image_list = listdir(f"{directory_root}/{plant_folder}/")
             `
make it like this since the listdir doesnot need addition folder in the end

**Comment by gshanbhag525** [comment]
> I need you to do following changes.
> 
> * Since plantdisease.zip after extraction becomes PlantVillage. Change line
> 
> `directory_root = 'PlantVillage'`
> 
> * `for plant_disease_folder in plant_disease_folder_list: print(f"[INFO] Processing {plant_disease_folder} ...") plant_disease_image_list = listdir(f"{directory_root}/{plant_folder}/") `
>   make it like this since the listdir doesnot need addition folder in the end

@AdityaSingh17

**Comment by AdityaSingh17** [comment]
@gshanbhag525 after unzipping the plantdisease.zip we get two folders plantvillage and PlantVillage both of them have deeper directories and hence I made the root directory as plantdisease.
However, as per your requested changed I have made the following change:
```
directory_root = 'plantdisease/plantvillage'
```

Addressing the second requested change, 
```
for plant_disease_folder in plant_disease_folder_list:
            print(f"[INFO] Processing {plant_disease_folder} ...")
            plant_disease_image_list = listdir(f"{directory_root}/{plant_folder}/{plant_disease_folder}")
```
The directory pointed by **{plant_disease_folder}** is required as we are interested in the images contained in the directory as well as the name of the directory (for labels). And hence, if we try to remove it, we will not be having access to the images listed any of the directories. So the second requested change is not possible without an extensive change of pre existing code.
However, let me know if anything more is needed from my side.

**Comment by gshanbhag525** [comment]
> @gshanbhag525 after unzipping the plantdisease.zip we get two folders plantvillage and PlantVillage both of them have deeper directories and hence I made the root directory as plantdisease.
> However, as per your requested changed I have made the following change:
> 
> ```
> directory_root = 'plantdisease/plantvillage'
> ```
> 
> Addressing the second requested change,
> 
> ```
> for plant_disease_folder in plant_disease_folder_list:
>             print(f"[INFO] Processing {plant_disease_folder} ...")
>             plant_disease_image_list = listdir(f"{directory_root}/{plant_folder}/{plant_disease_folder}")
> ```
> 
> The directory pointed by **{plant_disease_folder}** is required as we are interested in the images contained in the directory as well as the name of the directory (for labels). And hence, if we try to remove it, we will not be having access to the images listed any of the directories. So the second requested change is not possible without an extensive change of pre existing code.
> However, let me know if anything more is needed from my side.

Have you used this dataset?

**Comment by gshanbhag525** [comment]
@AdityaSingh17 When you download the dataset on your local system ....and unzip the datset it gives you only one folder 

> 'PlantVillage' 

. Just like the way it is shown in here [kaggle dataset](https://www.kaggle.com/emmarex/plantdisease)
Go at the end of webpage.

If you use google collab and import the kaggle datset there then it creates 2 folders after unizpping. 

1. PlantVillage
2. plantvillage/PlantVillage

Have a look here [Crop ai keras model](https://github.com/CropAi/Data-Modeling/blob/master/Crop_AI_Keras.ipynb)

**Comment by gshanbhag525** [comment]
> @gshanbhag525 after unzipping the plantdisease.zip we get two folders plantvillage and PlantVillage both of them have deeper directories and hence I made the root directory as plantdisease.
> However, as per your requested changed I have made the following change:
> 
> ```
> directory_root = 'plantdisease/plantvillage'
> ```
> 
> Addressing the second requested change,
> 
> ```
> for plant_disease_folder in plant_disease_folder_list:
>             print(f"[INFO] Processing {plant_disease_folder} ...")
>             plant_disease_image_list = listdir(f"{directory_root}/{plant_folder}/{plant_disease_folder}")
> ```
> 
> The directory pointed by **{plant_disease_folder}** is required as we are interested in the images contained in the directory as well as the name of the directory (for labels). And hence, if we try to remove it, we will not be having access to the images listed any of the directories. So the second requested change is not possible without an extensive change of pre existing code.
> However, let me know if anything more is needed from my side.

@AdityaSingh17 
On my machine the dataset it creates one folder ...In your system it creates 2 folders.
For the sake of simplicity and uniformity can you just create a collab jupyter notebook and import the dataset there .
For reference you can take a look at this notebook. [Crop AI keras model](https://github.com/CropAi/Data-Modeling/blob/master/Crop_AI_Keras.ipynb)

**Comment by AdityaSingh17** [comment]
In the notebook that you shared [here](https://github.com/CropAi/Data-Modeling/blob/master/Crop_AI_Keras.ipynb), notice there are two folders, plantvillage and PlantVillage.
The dataset is made in that way.
What exactly do you want me to do?
Screenshot from 2020-03-22 14-52-13

**Comment by AdityaSingh17** [comment]
Do you want me to shift my entire model architecture on colab?

**Comment by AdityaSingh17** [comment]
I have made the following changes to the code.
```
directory_root = 'PlantVillage'
```
and the processing function looks like:
```
image_list, label_list = [], []
try:
    print("[INFO] Loading images ...")
    root_dir = listdir(directory_root)
    for directory in root_dir :
        # remove .DS_Store from list
        if directory == ".DS_Store" :
            root_dir.remove(directory)

    for plant_disease_folder in root_dir:
        print(f"[INFO] Processing {plant_disease_folder} ...")
        plant_disease_image_list = listdir(f"{directory_root}/{plant_disease_folder}")
                
        for single_plant_disease_image in plant_disease_image_list :
            if single_plant_disease_image == ".DS_Store" :
                plant_disease_image_list.remove(single_plant_disease_image)

        for image in plant_disease_image_list[:200]:
            image_directory = f"{directory_root}/{plant_disease_folder}/{image}"
            if image_directory.endswith(".jpg") == True or image_directory.endswith(".JPG") == True:
                image_list.append(convert_image_to_array(image_directory))
                label_list.append(plant_disease_folder)
                    
    print("[INFO] Image loading completed!")  
except Exception as e:
    print(f"Error : {e}")
```
If you are satisfied with it, I will push the same.

**Comment by gshanbhag525** [comment]
it looks good..go ahead @AdityaSingh17

**Comment by AdityaSingh17** [comment]
@gshanbhag525 I have made the requested changes, please let me know if anything else is needed from my side!

**Comment by AdityaSingh17** [comment]
@gshanbhag525 request you to review the PR.
